data-prepper-plugins/kinesis-source/src/main/java/org/opensearch/dataprepper/plugins/kinesis/source/processor/KinesisRecordProcessor.java

```diff
@@ -87,87 +108,169 @@ private KinesisStreamConfig getStreamConfig(final KinesisSourceConfig kinesisSou
     public void initialize(InitializationInput initializationInput) {
         // Called once when the processor is initialized.
         kinesisShardId = initializationInput.shardId();
-        LOG.info("Initialize Processor for shard: {}", kinesisShardId);
+        String kinesisStreamName = streamIdentifier.streamName();
+        LOG.info("Initialize Processor for stream: {},  shard: {}", kinesisStreamName, kinesisShardId);
         lastCheckpointTimeInMillis = System.currentTimeMillis();
+
+        if (kinesisSourceConfig.isAcknowledgments()) {
+            executorService.submit(() -> monitorCheckpoint(executorService));
+        }
     }
 
-    private AcknowledgementSet createAcknowledgmentSet(final ProcessRecordsInput processRecordsInput) {
+    private void monitorCheckpoint(final ExecutorService executorService) {
+        while (!isStopRequested.get()) {
+            if (System.currentTimeMillis() - lastCheckpointTimeInMillis >= checkpointInterval.toMillis()) {
+                LOG.debug("Regular checkpointing for shard {}", kinesisShardId);
+
+                Optional<KinesisCheckpointerRecord> kinesisCheckpointerRecordOptional = kinesisCheckpointerTracker.getLatestAvailableCheckpointRecord();
+                if (kinesisCheckpointerRecordOptional.isPresent()) {
+                    RecordProcessorCheckpointer recordProcessorCheckpointer = kinesisCheckpointerRecordOptional.get().getCheckpointer();
```

Comment:
Can you combine these lines with 201-205?

Reply:
Thanks @dlvenable. I have refactored these lines and added a new method `doCheckpoint`.